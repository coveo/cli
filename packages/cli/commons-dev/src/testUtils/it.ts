import {fancy} from 'fancy-test';

/**
 * Wrapper around `jest.it` that captures all stdout/stderr.
 */
export const fancyIt = () => fancy.stdout().stderr().it;
