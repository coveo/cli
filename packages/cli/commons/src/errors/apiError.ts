import {red} from 'chalk';
import dedent from 'ts-dedent';
import {PrintableError, SeverityLevel} from './printableError';

export interface APIErrorResponse {
  message: string;
  errorCode: string;
  requestID: string;
}

export interface AxiosErrorFromAPI {
  response: {
    status: number;
    data: APIErrorResponse;
  };
}

export const APIErrorSchema = {
  type: 'object',
  required: ['errorCode', 'message'],
  properties: {
    errorCode: {
      type: 'string',
    },
    message: {
      type: 'string',
    },
  },
};

export const AxiosErrorFromAPISchema = {
  type: 'object',
  required: ['response'],
  properties: {
    response: {
      type: 'object',
      required: ['status', 'data'],
      properties: {
        status: {
          type: 'integer',
        },
        data: APIErrorSchema,
      },
    },
  },
};

export class APIError extends PrintableError {
  public constructor(
    error: APIErrorResponse | AxiosErrorFromAPI,
    tagLine?: string
  ) {
    super(SeverityLevel.Error);
    let errorCode: string;
    let message: string;
    let status: number | undefined;
    if (this.isFromAxios(error)) {
      errorCode = error.response.data.errorCode;
      message = error.response.data.message;
      status = error.response.status;
    } else {
      errorCode = error.errorCode;
      message = error.message;
    }

    this.name = `APIError - ${errorCode}`;
    this.message = dedent`
    ${tagLine}
    ${status ? `Status code: ${status}\n` : ''}
    Error code: ${red(errorCode)}
    Message: ${red(message)}
    `;
  }

  private isFromAxios(
    error: APIErrorResponse | AxiosErrorFromAPI
  ): error is AxiosErrorFromAPI {
    return 'response' in error;
  }
}
