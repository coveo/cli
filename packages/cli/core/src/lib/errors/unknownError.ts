import {
  PrintableError,
  SeverityLevel,
} from '@coveo/cli-commons/lib/errors/printableError';

export class UnknownError extends PrintableError {
  public name = 'Unknown CLI Error';
  public constructor(e?: unknown) {
    super(SeverityLevel.Error);
    const error = typeof e === 'string' ? new Error(e) : e;
    if (error && error instanceof Error) {
      this.message = error.message;
      this.stack = error.stack;
      this.name = `Unknown CLI Error - ${error.name}`;
    }
  }
}
