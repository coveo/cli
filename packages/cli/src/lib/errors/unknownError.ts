import {CLIBaseError} from './CLIBaseError';

export class UnknownError extends CLIBaseError {
  public name = 'Unknown CLI Error';
  public constructor() {
    super();
  }
}
