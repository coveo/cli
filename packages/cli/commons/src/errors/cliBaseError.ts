import {CLIError} from '@oclif/core/lib/errors';
import {Chalk, red, yellow, cyan} from 'chalk';

export enum SeverityLevel {
  Info = 'info',
  Warn = 'warn',
  Error = 'error',
}

export interface CLIBaseErrorInterface {
  level?: SeverityLevel;
  cause?: Error;
}

interface OClifCLIError extends Omit<CLIError, 'render' | 'bang' | 'oclif'> {}

export class CLIBaseError extends Error implements OClifCLIError {
  private static readonly defaultSeverity: SeverityLevel = SeverityLevel.Error;
  public name = 'CLI Error';

  public constructor(
    private error?: string | Error | CLIError,
    private options?: CLIBaseErrorInterface
  ) {
    super(error instanceof Error ? error.message : error, options);
  }

  public get stack(): string {
    return super.stack || '';
  }

  public get severityLevel(): SeverityLevel {
    return this.options?.level || CLIBaseError.defaultSeverity;
  }

  /**
   * Specific to internal oclif error handling
   */
  private get oclif() {
    return this.error instanceof CLIError ? this.error.oclif : {};
  }

  /**
   * Used by oclif to pretty print the error
   */
  private get bang() {
    let color: Chalk;

    switch (this.severityLevel) {
      case SeverityLevel.Info:
        color = cyan;
        break;

      case SeverityLevel.Warn:
        color = yellow;
        break;

      case SeverityLevel.Error:
      default:
        color = red;
        break;
    }
    return color('Â»');
  }
}
