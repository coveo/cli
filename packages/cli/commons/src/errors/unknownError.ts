import {CLIBaseError} from './cliBaseError';

export class UnknownError extends CLIBaseError {
  public name = 'Unknown CLI Error';
  public constructor(e?: unknown) {
    super(
      e instanceof Error
        ? {...e, name: `${UnknownError.name} - ${e.name}`}
        : undefined
    );
  }
}
