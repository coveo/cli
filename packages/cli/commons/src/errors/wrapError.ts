import {APIError, isAxiosError, isErrorFromAPI} from './apiError';
import {CLIBaseError} from './cliBaseError';
import {UnknownError} from './unknownError';

export function wrapError(err: unknown): CLIBaseError {
  if (typeof err === 'string') {
    return new CLIBaseError(err);
  }

  if (isAxiosError(err) || isErrorFromAPI(err)) {
    return new APIError(err);
  }

  if (err instanceof CLIBaseError) {
    return err;
  }

  if (err instanceof Error) {
    return new CLIBaseError(err);
  }

  return new UnknownError(err);
}
