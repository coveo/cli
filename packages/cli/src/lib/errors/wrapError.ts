import {APIError, isAxiosError, isErrorFromAPI} from './APIError';
import {CLIBaseError} from './CLIBaseError';
import {UnknownError} from './unknownError';

export function wrapError(err: unknown): CLIBaseError {
  if (typeof err === 'string') {
    return new CLIBaseError({message: err});
  }

  if (err instanceof CLIBaseError) {
    return err;
  }

  if (err instanceof Error) {
    return new CLIBaseError(err);
  }

  if (isAxiosError(err) || isErrorFromAPI(err)) {
    return new APIError(err);
  }

  return new UnknownError(err);
}
