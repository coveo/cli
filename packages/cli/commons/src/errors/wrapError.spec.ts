import stripAnsi from 'strip-ansi';
import {fancyIt} from '@coveo/cli-commons-dev/testUtils/it';
import {APIError} from './apiError';
import {CLIBaseError} from './cliBaseError';
import {UnknownError} from './unknownError';
import {wrapError} from './wrapError';

describe('wrapError', () => {
  let error: CLIBaseError;

  describe('when the error is a string', () => {
    beforeAll(() => {
      error = wrapError('this is an error');
    });

    fancyIt()('should instanciate a CLIBaseError', () => {
      expect(error).toBeInstanceOf(CLIBaseError);
    });

    fancyIt()('should store the message', () => {
      expect(error.message).toBe('this is an error');
    });
  });

  describe('when the error is coming from the API', () => {
    beforeAll(() => {
      const apiResponse = {
        message: 'Some error message',
        errorCode: 'SOMETHING_WENT_WRONG',
        requestID: 'some id',
      };
      error = wrapError(apiResponse);
    });

    fancyIt()('should instanciate an APIError', () => {
      expect(error).toBeInstanceOf(APIError);
    });

    fancyIt()('should prettify error message from API', () => {
      expect(stripAnsi(error.message)).toMatchSnapshot();
    });
  });

  describe('when the error is already a CLIBaseError', () => {
    const initialError = new CLIBaseError('ðŸ˜±');

    beforeAll(() => {
      error = wrapError(initialError);
    });

    fancyIt()('should return the same CLIBaseError instance', () => {
      expect(error).toBe(error);
    });
  });

  describe('when the error is a generic Error', () => {
    beforeAll(() => {
      const genericError = new Error('sad error ðŸ˜¥');
      error = wrapError(genericError);
    });

    fancyIt()('should instanciate a CLIBaseError', () => {
      expect(error).toBeInstanceOf(CLIBaseError);
    });

    fancyIt()('should persist error message', () => {
      expect(error.message).toBe('sad error ðŸ˜¥');
    });
  });

  describe('when the error is neither of the above', () => {
    const unknownError = {customParameter: 'foo'};

    beforeAll(() => {
      error = wrapError(unknownError);
    });

    fancyIt()('should instanciate an UnknownError', () => {
      expect(error).toBeInstanceOf(UnknownError);
    });
  });
});
