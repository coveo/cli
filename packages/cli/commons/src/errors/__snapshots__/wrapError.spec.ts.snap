// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapError when the error is coming from the API should prettify error message from API 1`] = `
"
Status code: 500
Error code: SOMETHING_WENT_WRONG
Message: Some error message
Request ID: some id"
`;
